fix(audio): set the bits-per-sample of Steam Streaming Speakers to 16-bit when the default audio device is 16-bit

andygrundman opened the pull request:
## Description
fix(audio): set the bits-per-sample of Steam Streaming Speakers to 16-bit when the default audio device is 16-bit. This may help with audio glitches when enabling the virtual device.


### Screenshot
<!--- Include screenshots if the changes are UI-related. --->


### Issues Fixed or Closed
<!--- Close issue example: `- Closes #1` --->
<!--- Fix bug issue example: `- Fixes #2` --->
<!--- Resolve issue example: `- Resolves #3` --->
Resolves https://github.com/orgs/LizardByte/discussions/682


## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

andygrundman commented:
I'm going to rewrite this as a method that returns the default device format and actually matches the SSS to the device. It doesn't need to be specific to 16 bits.

andygrundman commented:
This may have a bad bug :( I somehow caused it to run into some errors. I realized I'm not sure that I'm safely calling `default_device(device_enum)` and will need to read more of the docs. 

Info: Reinitializing audio capture
Error: Couldn't initialize audio client for [Stereo]: [0x8889000A]
Error: Couldn't find supported format for audio
Warning: Couldn't re-initialize audio input

ReenigneArcher commented on src/platform/windows/audio.cpp:
```diff
@@ -793,6 +795,25 @@ namespace platf::audio {
+        audio::prop_t prop;
+        prop_var_t current_device_format;
+
+        // clang-format off: avoid a long line and warnings about nested if statements
```
clang format is warning about nested if statements? I don't think I've seen that before... or do you mean sonar is?

andygrundman replied:
Sorry the if statements weren't clang warnings, just sonar. I'm still working on my C++ style, but I like the way this looks and it does clean up the if statements. (I also removed the one that was only for logging the device name). clang-format doesn't seem to have a line length limit and wanted to put this all on one line.

```
        if ( SUCCEEDED(current_default_dev->OpenPropertyStore(STGM_READ, &prop))
          && SUCCEEDED(prop->GetValue(PKEY_AudioEngine_DeviceFormat, &current_device_format.prop)) )
        {
```

Anyway, do you know if the API calls I'm making are safe? Or maybe the info I need can be found another way?

I forgot to mention one really strange bug I ran into. I use the S/PDIF device for stereo output and normally it's set to 24/48. I was switching to 16/48 for testing this. But at some point Windows must have re-selected it as default but chose the lowest (top-most) format in the list which is 16-bit 32kHz. I only noticed because my DAC displays the active samplerate and I have never seen it display 32 before. This must have happened during `reset_default_device()` but feels more like a Windows bug.

ReenigneArcher replied:
clang format does have a line length limit. Unfortunately it's not very good. When you set it, it not only makes sure that you don't exceed that limit, it also will make lines longer if you broke up it over several and it would fit within the limit. Due to this, I just set it to 0. In some cases I know you can separate it over multiple lines, but it probably depends on specifically what it is.

This is what everything looks like when set to 120. https://github.com/LizardByte/Sunshine/actions/runs/13682786643/job/38259008671?pr=3706

> Anyway, do you know if the API calls I'm making are safe? Or maybe the info I need can be found another way?

I'm not sure, not an expert in this area.

andygrundman commented:
OK, this is cleaned up and squashed to 1 commit. I discovered that the last patch I made to this file, removing f32 and s32 from stereo would break this new logic if you had a 32-bit default device. The correct fix for that previous issue is just to put these at the bottom of the list so they aren't preferred over 24 and 16.

This now works well for me and I can't reproduce the issue I had. I have also added extra logging that may prove helpful if this doesn't solve the issue for people. Specifically I log the bps and samplerate of the internal mixer format, with a note about auto-resampling if it's not 48k.

ReenigneArcher commented on src/platform/windows/audio.cpp:
```diff
@@ -117,19 +117,24 @@ namespace {
 
   using virtual_sink_waveformats_t = std::vector<WAVEFORMATEXTENSIBLE>;
 
+  // The list of virtual formats are sorted in preference order and the first valid format will be used.
+  // All bits-per-sample options are listed because we try to match this to the default audio device.
```
This seems like a good opportunity to document the template better. Could you document this template using a doxygen comment block?

```cpp
/*
 * @brief ...
 * @tparam ...
 */
```

- https://docs.lizardbyte.dev/projects/sunshine/latest/md_third-party_2doxyconfig_2docs_2source__code.html#documentation-blocks
- https://stackoverflow.com/a/13359876

ReenigneArcher commented on src/platform/windows/audio.cpp:
```diff
@@ -117,19 +117,24 @@ namespace {
     } else if (channel_count == 6) {
       auto channel_mask1 = waveformat_mask_surround51_with_backspeakers;
-      auto channel_mask2 = waveformat_mask_surround51_with_sidespeakers;
+      auto channel_mask2 = waveformat_mask_surround51_with_sidespeakers;  // XXX will never be used, but is probably the better 5.1 layout
```
What is `XXX` indicating here? Just calling extra attention to the note?

cgutman replied:
I'm not sure it's actually true either. I thought 5.1 with surrounds as back speakers was pretty universal. @andygrundman did you have more context to add here?

andygrundman replied:
Back in the DVD era, 5.1 had the 2 rear speakers on the side walls at 90 degrees and also they were a few feet above ear level. In the modern 5.1 layout for Atmos, Dolby says it's anywhere between 110 and 120 degrees, and everything is at ear level now. One thing that was never an actual thing is 5.1 with 180 degree rear wall speakers. Those speakers are what you add to the side speakers to get a 7.1 system (where sides are back to 90 degrees). So the right 5.1 would be putting them at 120, splitting the difference, but Microsoft is stuck with their legacy options, and an Atmos option that we can't really use.

I think in practice, it doesn't matter where the rear speakers are as long as it's sort of behind you a little bit. My comment preferring 5.1-side is partly because I'm old, but mostly referring to the way the ordering works in Sunshine, those other options won't ever be used and the user can't choose them. Really I'm happy if the channels are all lined up correctly on the client side. 7.1 was different from Windows in SDL for years and nobody noticed until we got it [patched](https://github.com/libsdl-org/SDL/issues/11709) recently.

<img height="200" alt="win-5 1-back" src="https://github.com/user-attachments/assets/690def9b-8bbf-4a57-b10f-2a3c4a95ee4a" />
<img height="200" alt="dolby-5 1" src="https://github.com/user-attachments/assets/22b3111f-986a-4b62-8035-e01ffe8a4b2d" />
<img height="200" alt="dolby-7 1" src="https://github.com/user-attachments/assets/194950d3-609d-40d6-aec3-3e5d50b9f4ac" />

ReenigneArcher commented on src/platform/windows/audio.cpp:
```diff
@@ -793,6 +802,26 @@ namespace platf::audio {
+        audio::prop_t prop;
+        prop_var_t current_device_format;
+
+        // clang-format off: easier to read this split over 2 lines
```
I know it's not ideal to have this on a long line, but I'd prefer not to have clang-format exceptions for this. I'm open to different clang-format rules though to help make things more readable.

massej commented:
My default device is 24bits and I need to set my streaming device to 16 bits. It would be nice to have an option in the GUI to force it to 16 bits. Everytime i start a game in moonlight my audio is crackling until I go under the device settings and set it to 16 bits.

massej commented:
It works, but I have to manually switch the settings every time, which gets a bit irritating.

After gaming, I need to switch the output back to 24-bit/384 kHz to enjoy music properly. If I leave it at 16-bit/48 kHz—my default setting for gaming—I can clearly notice a drop in audio quality. It's not an issue for games, but for music listening, it's far from ideal. (I'm using a Beyerdynamic headset with an audiophile-grade tube preamp.)

andygrundman commented:
Cool, so you obviously have a nice DAC. I used to think this was some sort of issue with bad motherboard audio, because I also use USB/SPDIF/HDMI audio and never have a problem with Steam speakers. And I always have my default audio set to 24/48. I guess you could test if 24/48 makes any difference to the problem. Oh, and did you check the Sunshine log?
